csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -11,130 +11,176 @@
             string adapterId,
             Current current,
             CancellationToken cancel) =>
-            _lookup.FindAdapterAsync(adapterId);
+            ResolveLocationAsync(new string[] { adapterId }, Protocol.Ice1, current, cancel);
 
         public ValueTask<IObjectPrx?> FindObjectByIdAsync(
             Identity id,
             Current current,
             CancellationToken cancel) =>
-            _lookup.FindObjectAsync(id);
+            ResolveWellKnownProxyAsync(id, Protocol.Ice1, current, cancel);
 
-        public ILocatorRegistryPrx? GetRegistry(Current current, CancellationToken cancel) => _registry;
+        public ValueTask<ILocatorRegistryPrx?> GetRegistryAsync(Current current, CancellationToken cancel) =>
+            new (_registry);
 
-        internal Locator(Lookup lookup, ILocatorRegistryPrx registry)
+        public ValueTask<IObjectPrx?> ResolveLocationAsync(
+            string[] location,
+            Protocol protocol,
+            Current current,
+            CancellationToken cancel) =>
+            _lookupServant.FindAdapterByIdAsync(location, protocol);
+
+        public ValueTask<IObjectPrx?> ResolveWellKnownProxyAsync(
+            Identity identity,
+            Protocol protocol,
+            Current current,
+            CancellationToken cancel) =>
+            _lookupServant.FindObjectByIdAsync(identity, protocol);
+
+        internal Locator(Lookup lookupServant, ILocatorRegistryPrx registry)
         {
-            _lookup = lookup;
+            _lookupServant = lookupServant;
             _registry = registry;
         }
     }
 
     internal class LocatorRegistry : ILocatorRegistry
     {
-        private readonly Dictionary<string, IObjectPrx> _adapters = new ();
+        private readonly Dictionary<(string, Protocol), IObjectPrx> _adapters = new ();
         private readonly object _mutex = new ();
-        private readonly Dictionary<string, HashSet<string>> _replicaGroups = new ();
+        private readonly Dictionary<(string, Protocol), HashSet<string>> _replicaGroups = new ();
 
-        public ValueTask SetAdapterDirectProxyAsync(
+        public void RegisterAdapterEndpoints(
             string adapterId,
-            IObjectPrx? proxy,
+            string replicaGroupId,
+            IObjectPrx endpoints,
             Current current,
             CancellationToken cancel)
         {
             lock (_mutex)
             {
-                if (proxy != null)
+                Protocol protocol = endpoints.Protocol;
+
+                if (replicaGroupId.Length > 0)
                 {
-                    _adapters[adapterId] = proxy.Clone(clearLocator: true, clearRouter: true);
+                    if (!_replicaGroups.TryGetValue((replicaGroupId, protocol), out HashSet<string>? adapterIds))
+                    {
+                        adapterIds = new HashSet<string>();
+                        _replicaGroups.Add((replicaGroupId, endpoints.Protocol), adapterIds);
```

Comment:
Use protocol instead of endpoints.Protocol

Reply:
Fixed.